### PR TITLE
[Refactor] CFA & SASM

### PR DESCRIPTION
**Pull request by zytx121**
https://github.com/open-mmlab/mmrotate/pull/523 is required.

- [x] Align test
- [x] Align train
- [x] UT

## Experiments
cfa_r50_fpn_1x_dota_qbox.py
report AP50: 69.63
AP50: 0.690415336526148
AP75: 0.4385321504041044
mAP: 0.41440689154197435

sasm_reppoints_r50_fpn_1x_dota_qbox.py
report AP50: 66.45
AP50: 0.6670036869881898
AP75: 0.4086403668534643
mAP: 0.3925785321392314

### CODE REVIEW

**Review by liuyanyi — APPROVED** [review]

File: mmrotate/models/dense_heads/cfa_head.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+                gt_instance. It usually includes ``bboxes`` and ``labels``
+                attributes.
+            batch_img_metas (list[dict]): Meta information of each image, e.g.,
+                image size, scaling factor, etc.
+            batch_gt_instances_ignore (list[:obj:`InstanceData`], optional):
+                Batch of gt_instances_ignore. It includes ``bboxes`` attribute
+                data that is ignored during training and testing.
+                Defaults to None.
+            stage (str): 'init' or 'refine'. Generate target for init stage or
+                refine stage.
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:273]:
```suggestion
                refine stage. Defaults to 'init'.
```

File: mmrotate/models/dense_heads/cfa_head.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+            unmap_outputs (bool): Whether to map outputs back to the original
+                set of anchors.
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:275]: default

File: mmrotate/models/dense_heads/cfa_head.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+            return_sampling_results (bool): Whether to return the sampling
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:276]:
```suggestion
            return_sampling_results (bool, optional): Whether to return the sampling
```

File: mmrotate/models/dense_heads/cfa_head.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+                - all_label_weights (list[Tensor]): Label weights of each
+                level.
+                - all_bbox_gt (list[Tensor]): Ground truth bbox of each level.
+                - all_proposals (list[Tensor]): Proposals(points/bboxes) of
+                each level.
+                - all_proposal_weights (list[Tensor]): Proposal weights of
+                each level.
+                - pos_inds (list[Tensor]): Index of positive samples in all
+                images.
+                - gt_inds (list[Tensor]): Index of ground truth bbox in all
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:292]:
```suggestion
            - all_labels (list[Tensor]): Labels of each level.
            - all_label_weights (list[Tensor]): Label weights of each
               level.
            - all_bbox_gt (list[Tensor]): Ground truth bbox of each level.
            - all_proposals (list[Tensor]): Proposals(points/bboxes) of
               each level.
            - all_proposal_weights (list[Tensor]): Proposal weights of
              each level.
            - pos_inds (list[Tensor]): Index of positive samples in all
              images.
            - gt_inds (list[Tensor]): Index of ground truth bbox in all
```
```suggestion
                - all_labels (list[Tensor]): Labels of each level.
                - all_label_weights (list[Tensor]): Label weights of each
                level.
                - all_bbox_gt (list[Tensor]): Ground truth bbox of each level.
                - all_proposals (list[Tensor]): Proposals(points/bboxes) of
                each level.
                - all_proposal_weights (list[Tensor]): Proposal weights of
                each level.
                - pos_inds (list[Tensor]): Index of positive samples in all
                images.
                - gt_inds (list[Tensor]): Index of ground truth bbox in all
```

File: mmrotate/models/dense_heads/cfa_head.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+            pos_mask = (0 <= single_labels) & (
+                single_labels < self.num_classes)
+            pos_inds.append(pos_mask.nonzero(as_tuple=False).view(-1))
+
+        gt_inds = [item.pos_assigned_gt_inds for item in sampling_result]
+
+        return (all_labels, all_label_weights, all_bbox_gt, all_proposals,
+                all_proposal_weights, pos_inds, gt_inds)
+
+    def get_pos_loss(self, cls_score, pts_pred, label, bbox_gt, label_weight,
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:329]: missing type hint

File: mmrotate/models/dense_heads/cfa_head.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+            pos_pts_pred,
+            pos_bbox_gt,
+            pos_convex_weight,
+            avg_factor=self.loss_cls.loss_weight,
+            reduction_override='none')
+        loss_cls = loss_cls.sum(-1)
+        pos_loss = loss_bbox + loss_cls
+        return pos_loss,
+
+    def reassign(self,
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:374]: missing typehint

File: mmrotate/models/dense_heads/cfa_head.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+            pos_gt_inds (Tensor): Gt_index of all positive samples got
+                from first assign process.
+            num_proposals_each_level (list, optional): Number of proposals
+                of each level.
+            num_level (int, optional): Number of level.
+
+        Returns:
+            tuple: Usually returns a tuple containing learning targets.
+
+                - label (Tensor): classification target of each anchor after
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/cfa_head.py:411]: format

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**